### PR TITLE
Reintroduce Using pdb2mdb from archive

### PR DESCRIPTION
**Pull request by alexischr**
This reverts commit e1762ea45a972f28bfbeccd804a788a627216927.

WIP

### CODE REVIEW

**Comment by alexischr** [comment]
@jonpryor @pjcollins Do I need the mono integration label to trigger the Windows build?

**Comment by pjcollins** [comment]
@alexischr no labels are needed for the new build pipeline, it will automatically run as part of the parent `Xamarin.Android` check. You can see the failed build here - https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=2735061&view=logs&jobId=cac0e8d3-0ef5-5d2b-b57e-e8fde7204df3. You can download more information from the `win-build-test-results` artifact found under the vertical ellipsis menu in the top right.

**Comment by alexischr** [comment]
While I was able to reproduce, I am still unable to track down the reason for the failure to load. The Assembly Binding Log Viewer on Windows actually claims that the load was successful:

```
*** Assembly Binder Log Entry  (6/6/2019 @ 10:32:05 PM) ***

The operation was successful.
Bind result: hr = 0x0. The operation completed successfully.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework\v4.0.30319\clr.dll
Running under executable  C:\Users\Administrator\Downloads\pdb2mdb-repo\pdb2mdb-repo\bin\Debug\pdb2mdb-repo.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = Mono.CompilerServices.SymbolWriter, Version=4.0.0.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756
 (Fully-specified)
LOG: Appbase = file:///C:/Users/Administrator/Downloads/pdb2mdb-repo/pdb2mdb-repo/bin/Debug/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = pdb2mdb-repo.exe
Calling assembly : pdb2mdb, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null.
===
LOG: This bind starts in default load context.
LOG: Using application configuration file: C:\Users\Administrator\Downloads\pdb2mdb-repo\pdb2mdb-repo\bin\Debug\pdb2mdb-repo.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework\v4.0.30319\config\machine.config.
LOG: Post-policy reference: Mono.CompilerServices.SymbolWriter, Version=4.0.0.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Users/Administrator/Downloads/pdb2mdb-repo/pdb2mdb-repo/bin/Debug/Mono.CompilerServices.SymbolWriter.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Users\Administrator\Downloads\pdb2mdb-repo\pdb2mdb-repo\bin\Debug\Mono.CompilerServices.SymbolWriter.dll
LOG: Entering run-from-source setup phase.
LOG: Assembly Name is: Mono.CompilerServices.SymbolWriter, Version=4.0.0.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756
LOG: Binding succeeds. Returns assembly from C:\Users\Administrator\Downloads\pdb2mdb-repo\pdb2mdb-repo\bin\Debug\Mono.CompilerServices.SymbolWriter.dll.
LOG: Assembly is loaded in default load context.
```

The versions, cultures and public key tokens match, so it is nothing obvious. @marek-safar maybe we know someone who can offer insight on Windows .NET (or we have a known issue?)

Again, the error is:
```
Could not load file or assembly 'Mono.CompilerServices.SymbolWriter, Version=4.0.0.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A)
```

**Comment by jonathanpeppers** [comment]
Here is a very tiny repro: [pdb2mdb-repo.zip](https://github.com/xamarin/xamarin-android/files/3267123/pdb2mdb-repo.zip)

If you run this, you'll get:

image

**Comment by alexischr** [comment]
@jonathanpeppers The above log is actually from the Fuslogvw.exe [0], as suggested by @akoeplinger, while running the pdb2mdb-repo solution.

Since then I found that adding the `x64` platform on the solution makes the reproduction go away - this only happens under `Any CPU` and `x86`

**Comment by jonathanpeppers** [comment]
Yeah, I just wanted to post the repro here, so it wasn't lost on Slack, thanks.

**Comment by dellis1972** [comment]
Looks like windows still has a problem 

```
warning : Could not load file or assembly 'Mono.CompilerServices.SymbolWriter, Version=4.0.0.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A) [E:\A\_work\2305\s\tests\Runtime-MultiDex\Mono.Android-TestsMultiDex.csproj]
E:\A\_work\2305\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2041,2): warning :    at Pdb2Mdb.Converter.Convert(String filename) [E:\A\_work\2305\s\tests\Runtime-MultiDex\Mono.Android-TestsMultiDex.csproj]
E:\A\_work\2305\s\bin\Release\lib\xamarin.android\xbuild\Xamarin\Android\Xamarin.Android.Common.targets(2041,2): warning :    at 
```

**Comment by dellis1972** [comment]
I suspect `Mono.CompilerServices.SymbolWriter` is using Cecil? In which case it would need to be remapped to use the `Xamarin.Android.Cecil.*` set of assemblies. We don't use `Mono.Cecil.*` directly because it can conflict with other versions loaded by other extensions in the IDE's. This can cause weird behaviour if a different version of Cecil is used. So we remap everything to use the `Xamarin.Android.Cecil.*` stuff using `remap-assembly-ref.exe`

**Comment by alexischr** [comment]
@jonathanpeppers `Mono.CompilerServices.SymbolWriter` does not use Cecil, so this is a return to the error we've had before. Also, downloading the appropriate Mono Android archive for this commit [0] we have an assembly that does not even have the previous problem that we noticed (Architecture is `AnyCPU (64-bit preferred)`. The reproduction zip that you included had a file where the Architecture was set to `32-bit`; that is either an intermittent bug with our bundle, or said remapping maybe somehow messes up strong name validation (note that we have `remap: true` for that assembly, and I think that's enough to trigger the `remap-assembly-ref.exe` tool)

I will try disabling remapping since it shouldn't be needed in this case. But other than that, I have found no hints of why this assembly fails at load (it verifies under both Mono and .NET `sn.exe -vf`)

[0] https://jenkins.mono-project.com/view/Per-commit%20packages/job/archive-mono/job/2019-02/280/Azure/processDownloadRequest/android-release-Windows-52f6ea149f412fa58a4d98a957fe85a2a7b720b5.zip

@jonathanpeppers @dellis1972 et al do we need a `mono integration` label for this build? I don't know if the assemblies are already remapped and stuck into the XA bundle, and so we need to rebuild the bundle to have one with a non-remapped assembly.

**Comment by alexischr** [comment]
There was an unrelated errors and I tried restarting the build, but now I see `Artifact prepare-build-results already exists for build 2801676.` I tried re-restarting.

**Comment by alexischr** [comment]
@grendello would you also consider the remapping to be a possible culprit for strong name validation failing on Windows .NET? There are no references to Cecil assemblies in `Mono.CompilerServices.SymbolWriter`, so setting `remap: false` on its `MonoUtilityFile` should be benign at worst, correct?

**Comment by grendello** [comment]
> @grendello Marek Habersack FTE would you also consider the remapping to be a possible culprit for strong name validation failing on Windows .NET? There are no references to Cecil assemblies in `Mono.CompilerServices.SymbolWriter`, so setting `remap: false` on its `MonoUtilityFile` should be benign at worst, correct?

Yes, it should be harmless.

**Comment by alexischr** [comment]
One more round, as pdb2mdb.exe itself does have a Mono.Cecil reference.

File: build-tools/xaprepare/xaprepare/ConfigAndData/Runtimes.cs [hunk 1]
```diff
@@ -389,6 +389,8 @@ partial class Runtimes
 			new MonoUtilityFile ("monodoc.dll.config",                     ignoreDebugInfo: true),
 			new MonoUtilityFile ("mono-api-html.exe",                      remap: true),
 			new MonoUtilityFile ("mono-api-info.exe",                      remap: true),
+			new MonoUtilityFile ("pdb2mdb.exe",                       	   remap: true),
```
Comment (jonathanpeppers) [inline build-tools/xaprepare/xaprepare/ConfigAndData/Runtimes.cs]: Looks like there are mixed spaces/tabs:
```suggestion
			new MonoUtilityFile ("pdb2mdb.exe",                            remap: true),
```

**Pull request merged**